python/taichi/lang/ast/ast_transformer.py

```diff
@@ -761,45 +762,73 @@ def build_Module(ctx, node):
         return None
 
     @staticmethod
-    def build_Attribute(ctx, node):
-        if node.attr == "append" and isinstance(node.value, ast.Subscript):
-            x = build_stmt(ctx, node.value.value)
-            if not isinstance(x, Field) or x.parent(
-            ).ptr.type != _ti_core.SNodeType.dynamic:
-                raise TaichiSyntaxError(
-                    f"In Taichi scope the `append` method is only defined for dynamic SNodes, but {x} is encountered"
-                )
-            index = build_stmt(ctx, node.value.slice)
-            node.value.ptr = None
-            node.ptr = lambda val: append(x.parent(), index, val)
+    def build_attribute_if_is_dynamic_snode_method(ctx, node):
+        is_subscript = isinstance(node.value, ast.Subscript)
+        names = ("append", "deactivate")
+        if node.attr not in names:
+            return False
+        if is_subscript:
+            x = node.value.value.ptr
+            indices = node.value.slice.ptr
+        else:
+            x = node.value.ptr
+            indices = []
+        if not isinstance(x, Field):
+            return False
+        if not x.parent().ptr.type == _ti_core.SNodeType.dynamic:
+            return False
+        snode = x._get_field_members()[0].ptr.snode()
+        field_dim = snode.num_active_indices()
```

Comment:
```suggestion
        field_dim = x.snode.ptr.num_active_indices()
```
Writing in this way might be cleaner :-)